File: Java/src/Construct.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.stream.Collector;
import java.util.stream.Collectors;

/*
    Verify if is possible to target word be constructed by concatenating elements of the string array.
*/

public class Construct {
    
    /**
     * ===============================================================
     *                         CAN CONSTRUCT           
     *                        decision problem                        
     * ===============================================================
     */

    /**
     * CAN CONSTRUCT NAIVE IMPLEMENTATION
     */
    public boolean naiveCanConstruct(String target, List<String> wordBank) {
        if (target.isEmpty()) { return true; }
        for (String subWord : wordBank) {
            if (!target.startsWith(subWord)) { continue; }
            if (naiveCanConstruct(target.replaceFirst(subWord, ""), wordBank)) {
                return true;
            } 
        }
        return false;
    }
    
    /**
     * CAN CONSTRUCT DYNAMIC IMPLEMENTATION
     * USING MEMOIZATION
     */
    public boolean dynamicCanConstructMemo(String target, List<String> wordBank) {
        return dynamicCanConstructMemo(target, wordBank, new HashMap<>());
    }
    private boolean dynamicCanConstructMemo(String target, List<String> wordBank, HashMap<String, Boolean> memo) {
        if (memo.containsKey(target)) { return memo.get(target); }
        if (target.isEmpty()) { return true; }
        for (String subWord : wordBank) {
            if (!target.startsWith(subWord)) { continue; }
            if (dynamicCanConstructMemo(target.replaceFirst(subWord, ""), wordBank, memo)) {
                memo.put(target, true);
                return true;
            }
        }
        memo.put(target, false);
        return false;
    }
    
    /**
     * CAN CONSTRUCT DYNAMIC IMPLEMENTATION
     * USING TABULATION
     */
    public boolean dynamicCanConstructTabular(String target, List<String> wordBank) {
        List<Boolean> tabulation = new ArrayList<>();
        tabulation.add(true);
        for (int i = 0; i < target.length(); i++) { tabulation.add(false); }

        for (int i = 0; i < target.length(); i++) {
            if (!tabulation.get(i)) { continue; }
            String word = target.substring(i);
            for (String prefix : wordBank) {
                if (word.startsWith(prefix)) {
                    tabulation.set(i + prefix.length(), true);
                }
                if (tabulation.get(target.length())) { return true; }
            }
        }
        return false;
    }
    
    /**
     * ===============================================================
     *                          COUNT CONSTRUCT                         
     *                       combinatoric problem                     
     * ===============================================================
     */

    public int naiveCountConstruct(String target, List<String> wordBank) {
        if (target.isEmpty()) { return 1; }
        int sum = 0;
        for (String prefix : wordBank) {
            if (!target.startsWith(prefix)) { continue; }
            sum += naiveCountConstruct(target.replaceFirst(prefix, ""), wordBank);
        }
        return sum;
    }

    /**
     * HOW COUNT DYNAMIC IMPLEMENTATION
     * USING MEMOIZATION
     */
    public Integer dynamicCountConstructMemo(String target, List<String> wordBank) {
        return dynamicCountConstructMemo(target, wordBank, new HashMap<>());
    }
    private Integer dynamicCountConstructMemo(String target, List<String> wordBank, HashMap<String, Integer> memo) {
        if (memo.containsKey(target)) { return memo.get(target); }
        if (target.isEmpty()) { return 1; }
        Integer sum = 0;
        for (String prefix : wordBank) {
            if (!target.startsWith(prefix)) { continue; }
            sum += dynamicCountConstructMemo(target.replaceFirst(prefix, ""), wordBank, memo);
        }
        memo.put(target, sum);
        return sum;
    }

    /**
     * HOW COUNT DYNAMIC IMPLEMENTATION
     * USING TABULATION
     */
    public Integer dynamicCountConstructTabular(String target, List<String> wordBank) {
        List<Integer> tabulation = new  ArrayList<>();
        tabulation.add(1);
        for (int i = 0; i < target.length(); i++) { tabulation.add(0); }

        for (int i = 0; i < target.length(); i++) {
            if (tabulation.get(i) == 0) { continue; }
            String word = target.substring(i);
            for (String prefix : wordBank) {
                if (word.startsWith(prefix)) {
                    tabulation.set(
                        i+prefix.length(),
                        tabulation.get(i) + tabulation.get(i+prefix.length())
                    );
                }
            }
        }
        return tabulation.get(target.length());
    }

    /**
     * ===============================================================
     *                        ALL CONSTRUCTS  
     *                      optimization problem                    
     * ===============================================================
     */

    /**
     * ALL CONSTRUCTS NAIVE IMPLEMENTATION
     */
    public List<List<String>> naiveAllConstruct(String target, List<String> wordBank) {
        if (target.isEmpty()) { return Arrays.asList(new ArrayList<>()); }
        List<List<String>> result = new ArrayList<>();
        for (String prefix : wordBank) {
            if (target.startsWith(prefix)) {
                List<List<String>> sufixWays =  naiveAllConstruct(target.replaceFirst(prefix, ""), wordBank);
                sufixWays.forEach( a -> a.add(0, prefix));
                result.addAll(sufixWays);
            }
        }
        return result;
    }

    /**
     * ALL CONSTRUCTS DYNAMIC IMPLEMENTATION
     * USING MEMOIZATION
     */
    public List<List<String>> dynamicAllConstructMemo(String target, List<String> wordBank) {
        return dynamicAllConstructMemo(target, wordBank, new HashMap<>());
    }
    private List<List<String>> dynamicAllConstructMemo(String target, List<String> wordBank, HashMap<String, List<List<String>>> memo) {
        if (memo.containsKey(target)) { return memo.get(target); }
        if (target.isEmpty()) { return Arrays.asList(new ArrayList<>()); }
        List<List<String>> result = new ArrayList<>();
        for (String prefix : wordBank) {
            if (target.startsWith(prefix)) {
                List<List<String>> sufixWays = dynamicAllConstructMemo(target.replaceFirst(prefix, ""), wordBank, memo)
                    .stream()
                    .map( a -> new ArrayList<>(a))
                    .collect(Collectors.toList());
                sufixWays.forEach( a -> a.add(0, prefix));
                result.addAll(sufixWays);
            }
        }
        memo.put(target, result);
        return result;
    }

    /**
     * BEST SUM DYNAMIC IMPLEMENTATION
     * USING TABULATION
     */
    public List<List<String>> dynamicAllConstructTabular(String target, List<String> wordBank) {
        List<List<List<String>>> tabulation = new ArrayList<>();
        tabulation.add(new ArrayList<>());
        tabulation.get(0).add(new ArrayList<>());
        for (int i = 0; i < target.length(); i++) { tabulation.add(new ArrayList<>()); }

        for (int i = 0; i <= target.length(); i++) {
            if (tabulation.get(i).isEmpty()) { continue; }
            String word = target.substring(i);
            for (String prefix : wordBank) {
                if (!word.startsWith(prefix)) { continue; }
                List<List<String>> current = tabulation.get(i)
                    .stream()
                    .map(a -> new ArrayList<>(a))
                    .collect(Collectors.toList());
                current.forEach(arr -> arr.add(prefix));
                tabulation.get(i+prefix.length()).addAll(current);
            }
        }
        return tabulation.get(target.length());
    }
    
}

File: Swift/Construct.swift
/**
* ===============================================================
*                         CAN CONSTRUCT           
*                        decision problem                        
* ===============================================================
*/

func naiveCanConstruct(target: String, wordBank: [String]) -> Bool {
    if target.isEmpty {
        return true
    }
    for word in wordBank {
        if target.hasPrefix(word) {
            var str = target
            str.removeFirst(word.count)
            if naiveCanConstruct(target: str, wordBank: wordBank) {
                return true 
            }
        }
    }
    return false
}

func dynamicMemoCanConstruct(target: String, wordBank: [String]) -> Bool {
    var memo: [String:Bool] = [:]
    func canConstruct(_ target: String, _ wordBank: [String], _ memo: inout [String:Bool]) -> Bool {
        if let value = memo[target] {
            return value
        }
        if target.isEmpty {
            return true
        }
        for word in wordBank {
            if !target.hasPrefix(word) {
                continue
            }
            var str = target
            str.removeFirst(word.count)
            if canConstruct(str, wordBank, &memo) {
                memo[target] = true
                return true
            }
        }
        memo[target] = false
        return false
    }
    return canConstruct(target, wordBank, &memo)
}

func dynamicTabularCanConstruct(target: String, wordBank: [String]) -> Bool {
    var tabulation = [true]
    for _ in 0 ..<  target.count {
        tabulation.append(false)
    }
    
    for i in 0 ..< target.count {
        if !tabulation[i] {
            continue
        }
        var str = target
        str.removeFirst(i)
        for word in wordBank {
            if str.hasPrefix(word) {
                tabulation[i + word.count] = true
            }
            if tabulation[target.count] {
                return true
            }
        }
    }
    return false
}

/**
* ===============================================================
*                          COUNT CONSTRUCT                         
*                       combinatoric problem                     
* ===============================================================
*/

extension String {
    func removeFirst(_ word: String) -> String {
        var str = self
        str.removeFirst(word.count)
        return str
    }
}

func naiveCountConstruct(target: String, wordBank: [String]) -> Int {
    if target.isEmpty { 
        return 1
    }
    var sum = 0
    for word in wordBank {
        if target.hasPrefix(word) {
            sum += naiveCountConstruct(target: target.removeFirst(word), wordBank: wordBank)
        }
    }
    return sum
}

func dynamicMemoCountConstruct(target: String, wordBank: [String]) -> Int {
    var memo: [String:Int] = [:]
    func countConstruct(_ target: String, _ wordBank: [String], _ memo: inout [String:Int]) -> Int {
        if let value = memo[target] {
            return value
        }
        if target.isEmpty {
            return 1
        }
        var sum = 0;
        for word in wordBank {
            if target.hasPrefix(word) {
                sum += countConstruct(target.removeFirst(word), wordBank, &memo)
            }
        }
        memo[target] = sum
        return sum
    }
    return countConstruct(target, wordBank, &memo)
}

func dynamicTabularCountConstruct(target: String, wordBank: [String]) -> Int {
    var tabulation: [Int] = [1]
    for _ in 0 ..< target.count {
        tabulation.append(0)
    }
    for i in 0 ..< target.count {
        if tabulation[i] == 0 {
            continue
        }
        var word = target
        word.removeFirst(i)
        for prefix in wordBank where word.hasPrefix(prefix) {
            tabulation[i+prefix.count] = tabulation[i] + tabulation[i+prefix.count]
        }
    }
    return tabulation[target.count]
}

/**
* ===============================================================
*                        ALL CONSTRUCTS  
*                      optimization problem                    
* ===============================================================
*/

func naiveAllConstruct(target: String, wordBank: [String]) -> [[String]] {
    if target.isEmpty {
        return [[]]
    }
    var result: [[String]] = []
    for prefix in wordBank where target.hasPrefix(prefix) {
        let sufixWays: [[String]] = naiveAllConstruct(target: target.removeFirst(prefix), wordBank: wordBank)
            .map { [prefix] + $0 }
        result += sufixWays
    }
    return result
}

func dynamicMemoAllConstruct(target: String, wordBank: [String]) -> [[String]] {
    var memo: [String:[[String]]] = [:]
    func allConstruct(_ target: String, _ wordBank: [String], _ memo: inout [String:[[String]]]) -> [[String]] {
        if let value = memo[target] {
            return value
        }
        if target.isEmpty {
            return [[]]
        }
        var result: [[String]] = []
        for prefix in  wordBank where target.hasPrefix(prefix) {
            let sufixWays = allConstruct(target.removeFirst(prefix), wordBank, &memo)
                .map { [prefix] + $0 }
            result += sufixWays
        }
        memo[target] = result
        return result
    }
    return allConstruct(target, wordBank, &memo)
}

func dynamicTabularAllConstruct(target: String, wordBank: [String]) -> [[String]] {
    var tabulation: [[[String]]] = [[[]]]
    for _ in 0 ..< target.count {
        tabulation.append([])
    }
    for i in 0 ... target.count where !tabulation[i].isEmpty {
        var word = target
        word.removeFirst(i)
        for prefix in wordBank where word.hasPrefix(prefix) {
            tabulation[i+prefix.count] += tabulation[i].map { $0 + [prefix] }
        }
    }
    return tabulation[target.count]
}
